invokeai/backend/util/devices.py

```diff
@@ -1,89 +1,120 @@
 
+# legacy APIs
+TorchPrecisionNames = Literal["float32", "float16", "bfloat16"]
 CPU_DEVICE = torch.device("cpu")
 CUDA_DEVICE = torch.device("cuda")
 MPS_DEVICE = torch.device("mps")
 
 
+def choose_precision(device: torch.device) -> TorchPrecisionNames:
+    """Return the string representation of the recommended torch device."""
+    warn(
+        "choose_precision() is deprecated. Use TorchDevice.choose_torch_dtype() instead.",
+        DeprecationWarning,
+        stacklevel=2,
+    )
+    torch_dtype = TorchDevice.choose_torch_dtype(device)
+    return PRECISION_TO_NAME[torch_dtype]
+
+
 def choose_torch_device() -> torch.device:
-    """Convenience routine for guessing which GPU device to run model on"""
-    config = get_config()
-    if config.device == "auto":
-        if torch.cuda.is_available():
-            return torch.device("cuda")
-        if hasattr(torch.backends, "mps") and torch.backends.mps.is_available():
-            return torch.device("mps")
-        else:
-            return CPU_DEVICE
-    else:
-        return torch.device(config.device)
-
-
-def get_torch_device_name() -> str:
-    device = choose_torch_device()
-    return torch.cuda.get_device_name(device) if device.type == "cuda" else device.type.upper()
-
-
-def choose_precision(device: torch.device) -> Literal["float32", "float16", "bfloat16"]:
-    """Return an appropriate precision for the given torch device."""
-    app_config = get_config()
-    if device.type == "cuda":
-        device_name = torch.cuda.get_device_name(device)
-        if "GeForce GTX 1660" in device_name or "GeForce GTX 1650" in device_name:
-            # These GPUs have limited support for float16
-            return "float32"
-        elif app_config.precision == "auto" or app_config.precision == "autocast":
-            # Default to float16 for CUDA devices
-            return "float16"
+    """Return the torch.device to use for accelerated inference."""
+    warn(
+        "choose_torch_device() is deprecated. Use TorchDevice.choose_torch_device() instead.",
+        DeprecationWarning,
+        stacklevel=2,
+    )
+    return TorchDevice.choose_torch_device()
+
+
+def torch_dtype(device: torch.device) -> torch.dtype:
+    """Return the torch precision for the recommended torch device."""
+    warn("torch_dtype() is deprecated. Use TorchDevice.choose_torch_dtype() instead.", DeprecationWarning, stacklevel=2)
+    return TorchDevice.choose_torch_dtype(device)
+
+
+NAME_TO_PRECISION: Dict[TorchPrecisionNames, torch.dtype] = {
+    "float32": torch.float32,
+    "float16": torch.float16,
+    "bfloat16": torch.bfloat16,
+}
+PRECISION_TO_NAME: Dict[torch.dtype, TorchPrecisionNames] = {v: k for k, v in NAME_TO_PRECISION.items()}
+
+
+class TorchDevice:
+    """Abstraction layer for torch devices."""
+
+    @classmethod
+    def choose_torch_device(cls, app_config: Optional[InvokeAIAppConfig] = None) -> torch.device:
```

Comment:
Just curious about what the `app_config` is doing here as an arg. Would you set `app_config.device` differently for each processor thread in the multi-GPU setup?

Same question for `choose_torch_dtype`

Reply:
Yes. It was part of a hack to make choose_torch_device() work in the multi-GPU context. I'm cleaning up the multi-GPU branch now and found a better way to handle this, so I have removed the argument from both calls.

Thanks for taking the time to review! I did not expect this PR to be such a pain.

Reply:
Got it. No worries :)